### ISSUE TITLE
perf(contrib): KMeans updates centroids with a per-cluster `nonzero` + `index_select` + `mean` loop (160 host syncs per fit at k=8); use scatter/bincount

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4533]
## Problem

`KMeans.fit` (`kornia/contrib/kmeans.py:180–186`) recomputes centroids as

```python
for index in range(self.num_clusters):
    selected = torch.nonzero(cluster_assignment == index).squeeze()
    selected = torch.index_select(X, 0, selected)
    ...
    current_centers[index] = selected.mean(dim=0)
```

`nonzero` is a host round trip on every accelerator (470 µs per call on MPS for 4k elements, #4522), so a fit with k=8 and 20 iterations does 160 of them plus 160 `index_select` and 160 `mean`. At 4096×32 the fit takes 29 ms on the M1 CPU and **94 ms on MPS**; at 65536×32 MPS wins (269 ms vs 460 ms) because the distance matrix dominates.

## Fix

One `index_add_` of `X` into a `(k, D)` zeros tensor by `cluster_assignment`, one `bincount(cluster_assignment, minlength=k)` for the counts, and a `where` on `count == 0` for the empty-cluster re-seed. No per-cluster loop, no sync per cluster; the empty-cluster random reseed can stay a masked `torch.randint` gather. (#2809 / #2861 already moved the distance to `kornia.geometry.linalg.euclidean_distance`; this is the other half.)

Measured on torch nightly 2.15.0.dev20260913, Apple M1 8 GB, macOS 26.5.1, kornia `633951bd7`, `PYTORCH_ENABLE_MPS_FALLBACK` unset, `torch.utils.benchmark` medians with `mps.synchronize` inside the timed region, fresh process per op. Split out of the MPS-audit tracking issue #4521; the PyTorch-side kernel findings are in #4522.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

**Comment by prithvirajrh** [comment]
Hi, I'd like to work on this one - vectorizing the centroid update with scatter/bincount to remove the per-cluster host syncs. Could you assign it to me?
